query/workmanager.go

```diff
@@ -0,0 +1,297 @@
+package query
+
+import (
+	"container/heap"
+	"errors"
+	"fmt"
+	"sync"
+)
+
+type batch struct {
+	queries []*Query
+	options *queryOptions
+	errChan chan error
+}
+
+// Config holds the configuration options for a new WorkManager.
+type Config struct {
+	// ConnectedPeers is a function that returns a channel where all
+	// connected peers will be sent. It is assumed that all current peers
+	// will be sent imemdiately, and new peers as they connect.
+	ConnectedPeers func() (<-chan Peer, func(), error)
+}
+
+// WorkManager is the main access point for outside callers, and satisfy the
+// QueryAccess API. It receives queries to pass to peers, and schedules them
+// among available workers, orchestrating where to send them.
+type WorkManager struct {
+	cfg *Config
+
+	// newBatches is a channel where new batches of queries will be sent to
+	// the workDispatcher.
+	newBatches chan *batch
+
+	// runWorker is function closure that should start a new worker for the
+	// given peer. We make this configurable to easily mock the worker used
+	// during tests.
+	runWorker func(peer Peer, nextJob <-chan *queryTask,
```

Comment:
Shouldn't the created worker be returned so we can manage its lifecycle externally?